dnscrypt.c

```diff
@@ -3,19 +3,29 @@
 const KeyPair *
 find_keypair(const struct context *c,
              const unsigned char magic_query[DNSCRYPT_MAGIC_HEADER_LEN],
-             const size_t dns_query_len)
+             const size_t dns_query_len,
+             bool *use_xchacha20)
 {
+    unsigned char magic_query_ref[DNSCRYPT_MAGIC_HEADER_LEN];
     const KeyPair *keypairs = c->keypairs;
     size_t i;
 
     if (dns_query_len <= DNSCRYPT_QUERY_HEADER_SIZE) {
         return NULL;
     }
+    *use_xchacha20 = 0;
     for (i = 0U; i < c->keypairs_count; i++) {
-        if (memcmp(keypairs[i].crypt_publickey, magic_query,
-                   DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+        memcpy(magic_query_ref, keypairs[i].crypt_publickey, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
             return &keypairs[i];
         }
+#ifdef HAVE_CRYPTO_BOX_CURVE25519XCHACHA20POLY1305_OPEN_EASY
+        sodium_increment(magic_query_ref, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+            *use_xchacha20 = 1;
```

Comment:
A `client_magic` represents a unique certificate at a given point in time.

That `client_magic` can be anything as long as it's not shared by two certificates. A new one must be issued when rotating keys. That's the only requirement.

`dnscrypt-wrapper` uses the first bytes of the public key to build the `client_magic`. 

Collisions may happen, but given the frequency of key rotations, they are actually unlikely to happen in practice, and this can be detected after creating a new key pair.